### PR TITLE
Fix empty editor windows

### LINKED ISSUE #7412: After closing Song Editor new project loaded ends with empty song editor window (must be closed and opened)
### System Information

Linux Debian Stable, wine32 on Linux, Windows 10 64-bit

### LMMS Version(s)

lmms-1.3.0-alpha.1.667+g1c865843f

### Most Recent Working Version

lmms-1.3.0-alpha.1.102+g89fc6c9-linux-x86_64.AppImage

### Bug Summary

At start working version (1.102) then compiled master (windows downloads - the same).
https://github.com/user-attachments/assets/b6bc879f-ca66-48ad-b8c0-ec69307458ef

The same bug is with Piano Roll, Pattern Editor, Automation Editor windows: if you will open project, saved in state (only) Piano Roll or Pattern Editor or Automation Editor window opened, just after this type window closed by mouse, this new project will be loaded with this type window in buggy state. 

### Expected Behaviour

If close Song Editor windows (with mouse or with Ctrl+W) after load another project song editor window should be not empty. Should not be needed to close and reopen it.  

### Steps To Reproduce

Load any project, close Song Editor window (as on video), open another project - project loading, but result - empty Song Editor window.

### Logs

<details>
<summary>Click to expand</summary>
<pre>
  
</pre>
</details>

### Screenshots / Minimum Reproducible Project

_No response_

### Please search the issue tracker for existing bug reports before submitting your own.

- [X] I have searched all existing issues and confirmed that this is not a duplicate.

### PR DESCRIPTION
**Pull request by firewall1110**
(1) Bug affects not only Song Editor window, but also Piano Roll, Pattern Editor, Automation Editor windows (all are extending Editor class). Using MainWindow::refocus() makes closing event handling close to closing by shortcut (from MainWindow code)
(2) This is BugFix PR so I do not change coding convention violation ( variable _ce ).
[Edited: after some refactoring commits by LMMS member I decide to change **_ce** to **event**]

Reworks #7035
Fixes #7412

### CODE REVIEW

**Comment by firewall1110** [comment]
@michaelgregorius  wrote
> ... with an explanation of why you think this fixes the bug. ...
> ensure that the solution does not break anything else what worked before

(1) I think that it fix the bug because I test my code. 

(2)  Bug is result of not proper another bug solution done by  PR #7035 

So I add another line 
`getGUI()->mainWindow()->refocus();`

but restore

`_ce->ignore();`

as it was before. 

I tested it against PR #7035 fixed bug steps:

_1: place notes on the grid.
2: reduce the grid until at least one of the notes doesn't line up.
3: play the loop.
4: close the piano roll editor.
5: hit space to pause._ 

so line 
`getGUI()->mainWindow()->refocus();`
is proper solution (and should be added in  PR #7035 instead to change `_ce->ignore();`). 

And this code is called, than windows are closed by Ctrl+1 , Ctrl+2, Ctrl+3, Ctrl+4 . 
So calling it from here can not harm more than calling in Ctrl+1 , Ctrl+2, Ctrl+3, Ctrl+4 context. 

(3)

Using `_ce->ignore();` in such situation is also in Qt 5.12 documentation:
> "The event handler [QWidget::closeEvent](https://hwzen.myds.me:17001/qtdoc/qtwidgets/qwidget.html#closeEvent)() receives close events. The default implementation of this event handler accepts the close event. If you do not want your widget to be hidden, or want some special handling, you should reimplement the event handler and [ignore](https://hwzen.myds.me:17001/qtdoc/qtcore/qevent.html#ignore)() the event."

**Review by irrenhaus3 — APPROVED** [review]
Fixes the bug. Might prefer a signal/slot channel instead of making the private `refocus()` public. Not a blocker in my eyes though.

**Comment by michaelgregorius** [comment]
Can someone explain why and how it fixes the bug? What caused the bug in the first place?

**Comment by firewall1110** [comment]
> how it fixes the bug?

Empty editors window bug is kind of bugs  I call "mystical bugs":
empty not functional window should not be fixed simply by hiding + un - hiding , but it happens , so all inner structures are ok. 
My guess is 
that after accepting close event Qt library make some kind of optimization , resource release, ... assuming, that this widget will not be used. So if we do not follow Qt documentation we get some undefined thing. 
But it is only my guess. 

> Might prefer a signal/slot channel instead of making the private refocus() public. 

My opinion is, that **signal/slot** channel is "overkill".
Variant is use **friend** line in Editor.h but:
 - if once it needed, may be will be needed once more;
 - from my point off view **friend** is more complex and more "exotic". 

> Commit https://github.com/LMMS/lmms/commit/4e78fb53dc945c4dd47b5d95279f7e9eb409df04 introduces modern C++ and slightly simplifies the logic.

It is better and is in LMMS style, so **I accept this**.

[ But I prefer old good (only for my taste illustration):
```
void MainWindow::refocus()
{
	QWidget* editor = getGUI()->songEditor()->parentWidget();
	if (!editor->isHidden()) { editor->setFocus(); return; }
	editor = getGUI()->patternEditor()->parentWidget();
	if (!editor->isHidden()) { editor->setFocus(); return; }
	editor = getGUI()->pianoRoll()->parentWidget();
	if (!editor->isHidden()) { editor->setFocus(); return; }
	editor = getGUI()->automationEditor()->parentWidget();
	if (!editor->isHidden()) { editor->setFocus(); return; }
	this->setFocus();
}
```
Modern compilers will make editor **register** variable, so it will be faster and more compact in result assembler code.

But such things not allowed here ... 
]

File: src/gui/MainWindow.cpp [hunk 1]
```diff
@@ -975,26 +975,23 @@ void MainWindow::toggleFullscreen()
...
-	{
-		if( ! (*editor)->isHidden() ) {
-			(*editor)->setFocus();
-			found = true;
-			break;
+	const auto gui = getGUI();
+
+	// Attempt to set the focus on the first of these editors that is not hidden...
+	const std::vector<QWidget*> editorParents = { gui->songEditor()->parentWidget(), gui->patternEditor()->parentWidget(),
+		gui->pianoRoll()->parentWidget(), gui->automationEditor()->parentWidget() };
```
Comment (messmerd) [inline src/gui/MainWindow.cpp:982]:
```suggestion
	const auto editorParents = std::array{ gui->songEditor()->parentWidget(), gui->patternEditor()->parentWidget(),
		gui->pianoRoll()->parentWidget(), gui->automationEditor()->parentWidget() };
```
There's a fixed number of elements so `std::vector` and its dynamic allocation is unnecessary

File: src/gui/MainWindow.cpp [hunk 2]
```diff
@@ -975,26 +975,23 @@ void MainWindow::toggleFullscreen()
...
+
+	for (auto editorParent : editorParents)
+	{
+		if(!editorParent->isHidden())
```
Comment (messmerd) [inline src/gui/MainWindow.cpp:986]:
```suggestion
		if (!editorParent->isHidden())
```
style

File: src/gui/MainWindow.cpp [hunk 3]
```diff
@@ -975,26 +975,23 @@ void MainWindow::toggleFullscreen()
...
+		{
+			editorParent->setFocus();
+			return;
 		}
 	}
 
-	if( ! found )
-		this->setFocus();
+	// ... otherwise set the focus on the main window.
+	this->setFocus();
```
Comment (messmerd) [inline src/gui/MainWindow.cpp:994]:
```suggestion
	this->setFocus();
```

**Comment by firewall1110** [comment]
Now in refocus() refactoring context I have one question:

_is **MainWindow** class is right place for **refocus()**?_ 

What about **GuiApplication** class?

P.S.
But my opinion : not here (in  this bug fix PR) such changes ...

**Comment by firewall1110** [comment]
Now MainWindow::refocus compiled with clang++ with exported CXXFLAGS="-Oz" is little bit more compact, than my "old style variant", but assembler listing is near the same ...

**Review by michaelgregorius — APPROVED** [review]
Approving this PR as it seems to fix the problem.

**Comment by firewall1110** [comment]
Thank You! 
Not so long ... ~ 15 days
P.S.
Should I open **_new issue_** about:
> is MainWindow class is right place for refocus()?
> What about GuiApplication class?

May be I find more things in MainWindow and Editor context to refactor.
Or it is **not actual**?

**Comment by michaelgregorius** [comment]
@firewall1110, I am currently not sure where to put it. I would have assumed that the other editors would be children of the `MainWindow` and in that case having `refocus` in `MainWindow` would be fitting as it would act on its children. However, in `GuiApplication` the editors seem to be parallel to `MainWindow` with regards to the hierarchy. It would be interesting to know if the windows are reparented at some point or what this all means.

So for now I'd rather keep the `refocus` method in `MainWindow` as it seems to make sense from an intuitive model.

**Pull request merged**